icu4c/source/common/uloc_tag.cpp

```diff
@@ -2093,11 +2093,13 @@ ultag_parse(const char* tag, int32_t tagLen, int32_t* parsedLen, UErrorCode* sta
             if (tagLen < newTagLength) {
                 uprv_free(tagBuf);
```

Comment:
It seems like this here is really the problematic line: Earlier on line 2065 `t` adopted tagBuf but t->buf is not cleared after this uprv_free(), so when the new tagBuf allocation fails, t->buf still has the pointer to the freed old buffer. The most natural fix seems to be to simply change the malloc line to start with `t->buf = tagBuf = [malloc...]` and not have a separate pointer-copy line. No comment needed.

**However**, I think a much more robust fix would be to change tagBuf from a raw pointer to a CharString, and to cloneData() its contents into t->buf only just before returning a successful `t`. That would also simplify appending string parts and NUL termination handling.

Also, I haven't looked at what's inside a ULanguageTag. An even bigger and potentially better fix might be to turn it into an internal C++ class LanguageTag that has a CharString buf field, and casting it to an opaque, C-style ULanguageTag at the C API boundaries. *This* would probably be fodder for a separate ticket.

Reply:
I agree w/ you changing tagBuf to CharString could be a more robust fix but since the logic here w/ tagBuf is very complicate in this function, a rewrite for such fix have more protential to introduce even more bugs than a simple local fix. I think we should focus on security fix on minimum risk to create more security bugs than focus on the local issue at hand. What you suggest is basically a rewrite that is a good task for a different reason/ticket to reduce the complexity and improve the robustness our code but it should not be a blood stopping fix for a security ticket.

Reply:
File ticket https://unicode-org.atlassian.net/browse/ICU-22476 for the CharString rewrite later so we will not forget about it.